src/transactions/contracts/HostVal.cpp

```diff
@@ -0,0 +1,676 @@
+    {
+        return ltOther(str);
+    }
+
+    bool
+    operator()(xdr::xvector<uint8_t> const& blob) const
+    {
+        return ltOther(blob);
+    }
+
+    bool
+    operator()(HostBigInt const& big) const
+    {
+        return ltOther(big);
+    }
+
+    bool
+    operator()(HostBigRat const& big) const
+    {
+        return ltOther(big);
+    }
+
+    bool
+    operator()(xdr::pointer<LedgerKey> const& lk) const
+    {
+        return ltOtherPtr(lk);
+    }
+
+    bool
+    operator()(xdr::pointer<Operation> const& op) const
+    {
+        return ltOtherPtr(op);
+    }
+
+    bool
+    operator()(xdr::pointer<OperationResult> const& ores) const
+    {
+        return ltOtherPtr(ores);
+    }
+
+    bool
+    operator()(xdr::pointer<Transaction> const& tx) const
+    {
+        return ltOtherPtr(tx);
+    }
+
+    bool
+    operator()(Asset const& a) const
+    {
+        return ltOther(a);
+    }
+
+    bool
+    operator()(Price const& p) const
+    {
+        return ltOther(p);
+    }
+
+    bool
+    operator()(AccountID const& a) const
+    {
+        return ltOther(a);
+    }
+};
+
+bool
+HostValInContext::operator<(HostValInContext const& other) const
+{
+    ZoneScoped;
+    if (mVal.isU63())
+    {
+        if (other.mVal.isU63())
+        {
+            return mVal.asU63() < other.mVal.asU63();
+        }
+        else
+        {
+            return true;
+        }
+    }
+    else if (other.mVal.isU63())
+    {
+        return false;
+    }
+    else if (mVal.getTag() != other.mVal.getTag())
+    {
+        return mVal.getTag() < other.mVal.getTag();
+    }
+
+    // We have the same tag, examine sub-cases.
+    if (mVal.isU32())
+    {
+        return mVal.asU32() < other.mVal.asU32();
+    }
+    else if (mVal.isI32())
+    {
+        return mVal.asI32() < other.mVal.asI32();
+    }
+    else if (mVal.isVoid())
+    {
+        return false;
+    }
+    else if (mVal.isBool())
+    {
+        return mVal.asBool() < other.mVal.asBool();
+    }
+    else if (mVal.isObject())
+    {
+        // Objects require deep comparison.
+        releaseAssert(mCtx);
+        auto const& ptr = mCtx->getObject(mVal);
+        auto const& otherPtr = mCtx->getObject(mVal);
+        if (!ptr && otherPtr)
+        {
+            return true;
+        }
+        else if (ptr && !otherPtr)
+        {
+            return false;
+        }
+        else if (!ptr && !otherPtr)
+        {
+            return false;
+        }
+        else
+        {
+            // Order by object-variant first.
+            auto idx = ptr->index();
+            auto otherIdx = ptr->index();
+            if (idx != otherIdx)
+            {
+                return idx < otherIdx;
+            }
+            // Deep-compare if they're the same variant.
+            LtHostObjectVisitor lhov{*otherPtr};
+            return std::visit(lhov, *ptr);
+        }
+    }
+    else if (mVal.isSymbol())
+    {
+        return mVal.asSymbol() < other.mVal.asSymbol();
```

Comment:
Going back to my comment about `asSymbol` skipping 0s, this seems problematic. We check if two symbols are equal by bitwise comparison. But we use their string representation to compare them for inequality. So it's possible that `a != b` but also `!(a < b) && !(b < a)` which seems less than ideal.